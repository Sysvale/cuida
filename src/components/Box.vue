<!-- eslint-disable vue/multi-word-component-names -->
<template>
	<div
		ref="cds-box"
		:class="elevated ? 'box--elevated' : `box--${variant}`"
	>
		<cds-clickable
			class="box__container"
			:clickable="clickable"
			@click="handleClick"
		>
			<!-- @slot Slot utilizado para renderização do conteúdo interno do Box.-->
			<cds-spacer
				:padding-top="padding"
				:padding-right="padding"
				:padding-bottom="padding"
				:padding-left="padding"
			>
				<slot />
			</cds-spacer>
		</cds-clickable>
	</div>
</template>

<script>
import CdsSpacer from './Spacer.vue';
import CdsClickable from './Clickable.vue';

import rounder from '../utils/methods/rounder';

export default {
	components: {
		CdsSpacer,
		CdsClickable,
	},

	props: {
		/**
		* Quando true aplica o token $shadow-sm como box-shadow do componente.
		*/
		elevated: {
			type: Boolean,
			default: false,
		},
		/**
		* Indica se o componente deverá ocupar 100% da largura disponível.
		*/
		fluid: {
			type: Boolean,
			default: false,
		},
		/**
		* O valor setado é sempre multiplicado por 4, assim como nos tokens scss do Cuida.
		* O valor mínimo aceito é 4 e o máximo aceito é 12.
		*/
		padding: {
			type: [Number, String],
			default: 4,
			validator: (value) => value <= 12,
		},
		/**
		* Ativa ou desativa o clique no componente
		*/
		clickable: {
			type: Boolean,
			default: false,
		},
		/**
		* A variante de cor. São 9 variantes implementadas: 'green', 'teal',
		* 'blue', 'indigo', 'violet', 'pink', 'red', 'orange', 'amber', 'gray' e 'dark'.
		*/
		variant: {
			type: String,
			default: 'gray',
		},
	},

	data() {
		return {
			width: 0,
		};
	},

	computed: {
		widthResolver() {
			return this.fluid ? '100%' : 'fit-content';
		},

		borderRadiusResolver() {
			return this.rounder(this.width, 16);
		},
	},

	mounted() {
		this.width = this.$refs['cds-box'].getBoundingClientRect().width;
	},

	methods: {
		rounder,

		handleClick() {
			if (this.clickable) {
				this.$emit('boxClick', true);
			}
		},
	},
}
</script>

<style lang="scss" scoped>
@import './../assets/sass/tokens.scss';

.box {
	width: v-bind(widthResolver);
	border-radius: v-bind(borderRadiusResolver);
	overflow: hidden;

	&__container {
		width: v-bind(widthResolver);
	}

	&--elevated {
		@extend .box;
		box-shadow: $shadow-sm;
		background-color: $n-0;
		outline: 1px solid $n-30;
	}

	@include variantResolver using ($color-name, $base-color, $disabled, $muted, $background, $hover) {
		@extend .box;
		outline: 1px solid $base-color;
		background-color: $background;
	}
}
</style>