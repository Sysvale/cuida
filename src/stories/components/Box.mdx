import { Meta, Story, Props, Controls, Canvas, Preview } from '@storybook/blocks';
import Box from '../../components/Box.vue';
import * as BoxStories from './Box.stories';

<Meta of={BoxStories} />

export const Template = (args) => ({
	components: { CdsBox: Box },
	setup() {
		return { args };
	},
	template: /*html*/ `
		<div style="height: 200px">
			<cds-box
				v-bind="args"
			>
				2
			</cds-box>
		</div>
	`,
	methods: {
		logClickableClick(event) {
			console.info('logClick: ', event);
		}
	},
});

# Box

### Boxes são componentes básicos de container que encapsulam conteúdo com o estilo padrão de borda, de sombra e de padding do cuida. Podem ser utilizados como componente primivo para cards, popovers, modais, etc.

***

<br />

## Preview

<Canvas>
  <Story of={BoxStories.Box} />
</Canvas>

<Controls />
