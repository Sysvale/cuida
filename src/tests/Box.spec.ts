// @vitest-environment jsdom
import { describe, test, expect } from 'vitest';
import Box from '../components/Box.vue';
import { mount } from '@vue/test-utils';

describe('Spacer snapshot test', () => {
	test('renders correctly', () => {
		const wrapper = mount(Box, {
			props: {
				elevated: false,
				fluid: false,
				padding: 4,
			},
			slots: {
				default:
					'<div style="width: 100px; height: 100px; background-color: #2AC092; border-radius: 20px"/>',
			},
		});

		expect(wrapper.html()).toMatchSnapshot();
	});
});
