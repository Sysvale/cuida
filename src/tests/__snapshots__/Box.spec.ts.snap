// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Spacer snapshot test > renders correctly 1`] = `
"<div data-v-f9290560="" class="box--gray">
  <!--
		Evento emitido quando o wrapper é clicado
		@event cds-click
		@type {Event}
	-->
  <div data-v-f9290560="" class="clickable__container--not-clickable box__container">
    <!-- @slot Slot utilizado para renderização do conteúdo interno do Box.-->
    <div data-v-9cc3ee57="" data-v-f9290560="" class="spacer">
      <div style="width: 100px; height: 100px; background-color: rgb(42, 192, 146); border-radius: 20px;"></div>
    </div>
  </div>
</div>"
`;
